src/Stratis.Bitcoin/P2P/Protocol/Behaviors/EnforcePeerVersionCheckBehavior.cs

```diff
@@ -0,0 +1,119 @@
+    {
+        /// <summary>An indexer that provides methods to query the best chain (the chain that is validated by the full consensus rules)</summary>
+        protected readonly ChainIndexer chainIndexer;
+
+        /// <summary>User defined node settings.</summary>
+        protected readonly NodeSettings nodeSettings;
+
+        /// <summary>Specification of the network the node runs on - regtest/testnet/mainnet.</summary>
+        protected readonly Network network;
+
+        /// <summary>Logger factory used while cloning the object.</summary>
+        protected readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        protected readonly ILogger logger;
+
+        /// <summary>
+        /// Set to <c>true</c> if the attached peer callbacks have been registered and they should be unregistered,
+        /// <c>false</c> if the callbacks are not registered.
+        /// </summary>
+        protected bool callbacksRegistered;
+
+        /// <summary>
+        /// Initializes an instance of the object for outbound network peers.
+        /// </summary>
+        /// <param name="chainIndexer">The chain of blocks.</param>
+        /// <param name="nodeSettings">User defined node settings.</param>
+        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public EnforcePeerVersionCheckBehavior(ChainIndexer chainIndexer,
+            NodeSettings nodeSettings,
+            Network network,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainIndexer = chainIndexer;
+            this.nodeSettings = nodeSettings;
+            this.network = network;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        [NoTrace]
+        protected Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            int enforceMinProtocolVersionAtBlockHeight = this.network.Consensus.Options.EnforceMinProtocolVersionAtBlockHeight;
+            bool enforcementRequired = enforceMinProtocolVersionAtBlockHeight > 0;
+            if (!enforcementRequired)
+            {
+                this.logger.LogTrace("(-)[ENFORCEMENT_HEIGHT_NOT_REACHED]");
+                return Task.CompletedTask;
+            }
+
+            bool enforcedAlready = this.nodeSettings.MinProtocolVersion >= this.network.Consensus.Options.EnforcedMinProtocolVersion;
+            if (enforcedAlready)
+            {
+                this.logger.LogTrace("(-)[ALREADY_ENFORCED]");
+                return Task.CompletedTask;
+            }
+
+            bool enforcementHeightReached = this.chainIndexer.Height >= enforceMinProtocolVersionAtBlockHeight;
+            if (enforcementHeightReached)
```

Comment:
I don't think this `if` condition is needed, the behaviour will be executed for all peers so changing `this.nodeSettings.MinProtocolVersion` should not matter (as you disconnect the peer in the next check)

Reply:
The if statement ensures that the log Message about changing the min supported version is logged only once. I'm going to change this to Info level.

Reply:
ok yeah my issue is not the `if` itself but the change to the protocol settings, its not such a good practice settings should remain defined by config.